### DISCUSSION TITLE
Use tree-sitter in Rust, compile everything to WASM?

### DISCUSSION THREAD

**Discussion by jrieken** [tree-sitter/tree-sitter#1550] [Q&A] [answered]
Assume I have a library that is written in Rust and that uses tree-sitter via its rust bindings. Next, I want to compile that library to WASM and expose its features via `wasm_bindgen`.  

As a sample I have the snippet below and I try to compile it via `cargo b --target wasm32-unknown-unknown` (without success). Did anyone try this? Should this be asked in some Rust forum or somewhere else?

```rs
use tree_sitter::Parser;
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub fn foo() -> u64 {
    Parser::new().timeout_micros()
}

```

Published a simple sample here: https://github.com/jrieken/rust-wasm (run cargo b --target wasm32-unknown-unknown, easiest with [dev containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers))

/cc @dcreager

**Comment by liamrosenfeld** [comment]
The ABI incompatibility between Clang and Rust for wasm-unknown-unknown causes issues for this. The easiest way around that currently is to use the Rust port of tree sitter here: https://github.com/shadaj/tree-sitter-c2rust/tree/c2rust-master

**Reply by kamoshi** [reply]
Is there a similar thing available for Treesitter Highlight? It would be useful to be able to run Treesitter highlighting in a wasm setting

**Reply by liamrosenfeld** [reply]
I think that project also includes a translated tree sitter highlight. I haven't tested it though since what I'm working on involves highlighting already parsed trees which is not a feature of normal tree sitter highlight. So I'm using a modified version that just lives in a file in my project.

**Reply by mainrs** [reply]
That only works for the base `tree-sitter` package. As soon as one wants to include a language, like [Julia](https://github.com/tree-sitter/tree-sitter-julia/blob/master/Cargo.toml), the compilation fails again.

Can a similar approach be used for languages I want to use inside of WASM?

**Comment by kamoshi** [comment]
I've tried to compile `tree-sitter`, `tree-sitter-highlight`, and some languages, via rustc to `wasm-unknown-unknown`, however the problem is that the source code for tree-sitter contains `#include <stdio.h>`. This header is not available while compiling for `wasm-unknown-unknown`, and it can only be compiled with emscripten, which is a quite problematic target in rustc

**Reply by liamrosenfeld** [reply]
I got around that issue for the grammars by providing what was missing through shim files that are added when compiling dependencies though environment variables (`wasm-sysroot/` and `justfile` here: https://github.com/cacticouncil/lilypad).

However, if you're using tree-sitter from rust for `wasm32-unknown-unknown`, you're gonna run into the ABI incompatibility between rust and C. You can join the club of waiting on this to merge to finally fix it: https://github.com/rust-lang/rust/pull/117919

**Comment by benwis** [comment]
@liamrosenfeld I was hoping you could tell me a bit more about what you meant in the comment above. I too want to compile treesiter into a wasm file that can be loaded on the web. I'm confused about what you meant by using tree-sitter from rust for wasm32-unknown-unknown. Sorry for bugging you!

**Reply by liamrosenfeld** [reply]
If you just want to use wasm compiled tree sitter in the browser from javascript, you can follow [this](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/README.md). What I've said here is about compiling a Rust project that depends on tree sitter to web assembly (there is currently an issue with the Rust compiler that requires some workarounds when doing so).

**Reply by mainrs** [reply]
This has been merged some days ago. Is there another issue that I can track or get involved in to add direct compilation to Rust?

**Reply by liamrosenfeld** [reply]
Yeah now that https://github.com/rust-lang/rust/pull/117919 has merged, you can build it on nightly. However, there is a small import issue in tree sitter that my PR here fixes https://github.com/tree-sitter/tree-sitter/pull/3304. When using that fix, the only thing that is needed is a shim C standard library (like an extended version of https://github.com/cacticouncil/lilypad/tree/main/wasm-sysroot)

**Reply by nicx256** [reply]
if anyone else wants to go this route, i've muddled together [the required libc shim here](https://gist.github.com/nicx256/c86f172cece6370b6d503fa440473896) (nb: amateur code. it's probably bad, but it works)

**Reply by agentcooper** [reply]
@nicx256 @liamrosenfeld I tried a combination of your approaches (see test repo https://github.com/agentcooper/tree-sitter-wasm-demo), but getting compilation errors with `rustc 1.82.0-nightly` and `wasm-c-abi=spec`:
```
warning: tree-sitter@0.22.6: In file included from /Users/agentcooper/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tree-sitter-0.22.6/src/lib.c:3:
warning: tree-sitter@0.22.6: In file included from /Users/agentcooper/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tree-sitter-0.22.6/src/./alloc.c:1:
warning: tree-sitter@0.22.6: /Users/agentcooper/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tree-sitter-0.22.6/src/./alloc.h:18:45: error: a parameter list without types is only allowed in a function definition
warning: tree-sitter@0.22.6:    18 | TS_PUBLIC extern void *(*ts_current_malloc)(size_t);
```
Maybe you have any ideas how to fix it?

**Reply by nicx256** [reply]
not at the moment, sorry: i successfully built your demo repo using 1.82.0-nightly 2024-08-18, tree-sitter 0.22.6 and my Manjaro Linux system's include files, i.e. `CFLAGS_wasm32_unknown_unknown="-I/usr/include"` on the command line, so the problem's possibly with the `wasm-sysroot` thing you're using (iirc, i couldn't get that method to work either, i think the person who suggested it was using macOS?)

i tried building on a mac machine using the invocation in your readme, and got the same error

i also tried building on a mac using the output of `xcrun --show-sdk-path` which should be the location of Xcode's C header files, i.e.

```
CFLAGS_wasm32_unknown_unknown="-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/" RUSTFLAGS="-Zwasm-c-abi=spec" wasm-pack build --target web --dev
```
..but got a different error: "error: architecture not supported"

**Reply by liamrosenfeld** [reply]
I was able to get it working here: https://github.com/cacticouncil/lilypad

Relevant files: `c_shim.rs`, `wasm-sysroot/*`, `justfile`

**Comment by mainrs** [comment]
> Yeah now that [rust-lang/rust#117919](https://github.com/rust-lang/rust/pull/117919) has merged, you can build it on nightly. However, there is a small import issue in tree sitter that my PR here fixes #3304. When using that fix, the only thing that is needed is a shim C standard library (like an extended version of [cacticouncil/lilypad@`main`/wasm-sysroot](https://github.com/cacticouncil/lilypad/tree/main/wasm-sysroot))

Would this also work for `tree-sitter-lang` packages? I tried patching them with a custom `patch.crates-io` entry, but that didn't work. Maybe the shim was missing. It couldn't find `stdlib.h`.

**Reply by liamrosenfeld** [reply]
Currently, to provide the shim to the cargo dependencies, I'm setting `CFLAGS_wasm32_unknown_unknown := echo "-I$(pwd)/wasm-sysroot"` at build time.

**Comment by benwis** [comment]
It's a bit different, but I was able to compile treesitter and its parsers as part of my markdown to html compiler crate femark to wasm via the wasi target, and deploy in the browser. I wrote about that process here:  https://benw.is/posts/compiling-rust-to-wasi

**Reply by edusporto** [reply]
Hi benwis, I was following your steps and trying to compile tree-sitter into WASI, but I got the following compilation error when running the same command as you described (updated to the newer wasm-sdk version):
```
> RUSTFLAGS='-L /opt/wasi-sdk-24.0/share/wasi-sysroot/lib/wasm32-wasi -lstatic=c++ -lstatic=c++abi' CXXSTDLIB=c++ CC=/opt/wasi-sdk-24.0/bin/clang CXX=/opt/wasi-sdk-24.0/bin/clang++ CXXFLAGS="-fno-exceptions" cargo component build --release

...

warning: tree-sitter@0.22.6: In file included from /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tree-sitter-0.22.6/src/lib.c:13:
warning: tree-sitter@0.22.6: /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tree-sitter-0.22.6/src/./tree.c:156:10: error: call to undeclared function 'dup'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
warning: tree-sitter@0.22.6:   156 |   return dup(file_descriptor);
warning: tree-sitter@0.22.6:       |          ^
warning: tree-sitter@0.22.6: 1 error generated.

error: failed to run custom build command for `tree-sitter v0.22.6`
```
It seems like WASI can't find the `dup` function from `unistd.h`. Did you have this problem back when you wrote the article, or is it something new? I'm not sure if this problem stems from a new version of tree-sitter, a new version of wasm-sdk, or just some misconfiguration from my part.

**Reply by agentcooper** [reply]
@edusporto It seems WASI does not support `dup`, but compiled code worked fine for my use case after compiling with `CFLAGS="-Wno-everything"`.

Full command I used:
```
RUSTFLAGS="-L /Users/agentcooper/Downloads/wasi-sdk-24.0-arm64-macos/share/wasi-sysroot/lib/wasm32-wasip2" CC="/Users/agentcooper/Downloads/wasi-sdk-24.0-arm64-macos/bin/clang" CFLAGS="-Wno-everything" cargo build --target=wasm32-wasip2
```

**Reply by edusporto** [reply]
@agentcooper thank you very much, that did compile! Now I'll see if my binary works as intended :P

**Comment by StratusFearMe21** [comment]
I had luck with using [this](https://github.com/shadaj/tree-sitter-c2rust/tree/c2rust-master) version of `tree-sitter`, and putting a `tree-sitter` generated parser in the path of the `c2rust` transpiler, and linking it directly to the `tree-sitter` library. However, for the parser I'm compiling, it takes a comically long time to transpile (around 10 minutes on my machine), and also to compile the generated rust code. The parser code I put in came out as an 8MB rust file, so the long compile time definitely scans.

**Reply by mainrs** [reply]
So you only use the c2rust crate version to transpile the parsers you want from C to Rust? But at the end, you link against the normal `tree-sitter` Rust bindings? 

I made it work with c2rust too, but as soon as I load my WASM, I get a stack overflow. It happens whenever I create a new Parser using `Parser::new()`. I couldn't figure out why...

**Reply by StratusFearMe21** [reply]
For me, I got a Chrome Aw, Snap! Error. It was fixed by compiling in release mode

**Reply by mainrs** [reply]
Could you explain to me again how you compiled it in detail? Or just point to a repository that contains the code? Because for me WASM keeps producing these crashes.

**Reply by StratusFearMe21** [reply]
Yeah. Ill post the code here later today

**Reply by StratusFearMe21** [reply]
@mainrs Here is the source code and instructions for getting tree-sitter parsers working under Rust WASM. Enjoy :smile: !

https://github.com/StratusFearMe21/tree-sitter

**Reply by StratusFearMe21** [reply]
@mainrs Fixed an error in the README (if you've already gotten started with the repo)

**Reply by mainrs** [reply]
Thanks! This only works as long as I want to include a single language, right? If I want to use it with Java and Haskell, then this method would no work.

**Reply by StratusFearMe21** [reply]
The only thing stopping you from including more languages is your computer's RAM. `c2rust` will run out of memory if you ask it to transpile too many parsers

Say we want to include another `tree-sitter` parser, and we put it in `lib/src/another_parser.c`. These are all the changes you would need to make to the repo to include that new parser. From there, the rest of the instructions in the repo will apply like normal

```diff
diff --git a/lib/binding_rust/core_wrapper/mod.rs b/lib/binding_rust/core_wrapper/mod.rs
index c31e9899..9c0a34fd 100644
--- a/lib/binding_rust/core_wrapper/mod.rs
+++ b/lib/binding_rust/core_wrapper/mod.rs
@@ -17,6 +17,8 @@ pub mod core {
 
     pub use c2rust_bitfields::*;
 
+    #[allow(unused, clippy::all)]
+    pub(crate) mod another_parser;
     #[allow(unused, clippy::all)]
     pub(crate) mod api_raw;
     #[allow(unused, clippy::all)]
@@ -105,6 +107,7 @@ pub mod core {
 
     pub use alloc::*;
 
+    pub use another_parser::*;
     pub use api_raw::*;
     pub use get_changed_ranges::*;
     pub use imported_parser::*;
diff --git a/lib/src/lib.c b/lib/src/lib.c
index f9d15740..0fda74e6 100644
--- a/lib/src/lib.c
+++ b/lib/src/lib.c
@@ -1,6 +1,7 @@
 #define _POSIX_C_SOURCE 200112L
 
 #include "./alloc.c"
+#include "./another_parser.c"
 #include "./get_changed_ranges.c"
 #include "./imported_parser.c"
 #include "./language.c"
diff --git a/meson.build b/meson.build
index 68352513..18120da8 100644
--- a/meson.build
+++ b/meson.build
@@ -16,6 +16,7 @@ tree_sitter_lib = library(
   'lib/src/processed/tree_cursor.c',
   'lib/src/processed/wasm_store.c',
   'lib/src/processed/imported_parser.c',
+  'lib/src/processed/another_parser.c',
   include_directories: ['lib/include', 'lib/src', 'lib/src/wasm']
 )
 
diff --git a/process_c2rust/src/main.rs b/process_c2rust/src/main.rs
index 6b51712c..48cb30f4 100644
--- a/process_c2rust/src/main.rs
+++ b/process_c2rust/src/main.rs
@@ -525,6 +525,7 @@ fn main() {
     );
     public_apis.insert("lib/src/wasm_store.rs", vec!["TSWasmStore"]);
     public_apis.insert("lib/src/imported_parser.rs", vec![]);
+    public_apis.insert("lib/src/another_parser.rs", vec![]);
 
     public_apis
         .keys()
```

**Reply by nicx256** [reply]
@StratusFearMe21 Thanks for the example code and instructions, unfortunately step d (running mason) fails with "meson.build:3:18: ERROR: File lib/src/processed/lib.c does not exist."

**Reply by StratusFearMe21** [reply]
@nicx256 oops, try running `transpile.sh`, then running `meson`, then continuing the instructions from there.

**Reply by nicx256** [reply]
@StratusFearMe21 in the end i couldn't manage to build c2rust version, but wrote [a libc shim here](https://gist.github.com/nicx256/c86f172cece6370b6d503fa440473896) you can use with straight clang-compiled-for-wasm tree sitter

**Reply by hediet** [reply]
@nicx256 Rust/c noob here. How can your libc shim be used to get tree sitter to work with rust/wasm?

**Reply by nicx256** [reply]
@hediet This is how i do it on linux, liamrosenfeld's [approach above](https://github.com/tree-sitter/tree-sitter/discussions/1550#discussioncomment-9190112) of including his own copy of the libc headers will work on other OSes that don't put them in `/usr/include/`.

Put the shim file `wasm_libc.rs` in your `src` directory, say `mod wasm_libc;` in your `lib.rs`, and compile using the command 

```CFLAGS_wasm32_unknown_unknown="-I/usr/include" RUSTFLAGS="-Zwasm-c-abi=spec" wasm-pack build --target web --dev```

..you need the -I include flag to tell `tree-sitter` crate where the header files are, and to be compiling with rustc nightly so the `-Zwasm-c-abi=spec` flag works.

**Comment by agentcooper** [comment]
Here is an example setup for compiling to WebAssembly and running on web through WASI: https://github.com/agentcooper/tree-sitter-wasi-web.

**Reply by hediet** [reply]
Nice, thanks for the sample!

**Reply by michaelangeloio** [reply]
@agentcooper question, do you think this approach will work for most languages?

**Comment by Cypher1** [comment]
Imo it'd be really great to have a pure rust implementation of tree sitter lang code so that these problems can be avoided completely.

Is there any work to do this or is there a better work around?

**Comment by ObserverOfTime** [comment, marked as answer]
Implemented in #4820

**Comment by 80avin** [comment]
I was stuck around it while trying to use tree-sitter and tree-sitter-julia in a Dioxus (web) app.

I needed:
* https://github.com/cacticouncil/lilypad/tree/main/wasm-sysroot
* nightly toolchain ( 1.94.0-nightly ). Add following to `rust-toolchain.toml`
   ```toml
   [toolchain]
   channel = "nightly"
   ```
* Start using:
   ```sh
   CFLAGS_wasm32_unknown_unknown="-I$(shell pwd)/wasm-sysroot -Wbad-function-cast -Wcast-function-type -fno-builtin" dx serve --platform web
   ```
